File: CSC330/poker/Go/Poker.go
package main
import (
	"fmt"
	"os"
	"bufio"
	"strings"
)

func checkForDuplicates(card Card, dealtCards map[int]bool) {
	cardKey := card.getRank() * 4 + card.getSuit() // Combine rank and suit to form a unique key
	if dealtCards[cardKey] {
		panic(fmt.Sprintf("Error: Duplicate card detected: %s%s", card.rank, card.suit))
	}
	dealtCards[cardKey] = true
}
func main() {
// ======================= Set up ===========================
        fmt.Println("*** P O K E R  H A N D  A N A L Y Z E R ***\n\n")
	// Create 6 hands
	hands := make([]*Hand, 6)
	for i := range hands {
		hands[i] = NewHand()
	}
	if(len(os.Args) > 1){
// ================= Using Test Deck =======================
		dealtCards := make(map[int]bool)
		currentHand := 0
		filename := os.Args[1]
		fmt.Printf("*** File: %s\n\n", filename)
		// Open the file
		file, err := os.Open(filename)
		if err != nil {
			fmt.Printf("Error opening file: %v\n", err)
			return
		}
		defer file.Close()

		// Read all from file
		var allLines []string
		scanner := bufio.NewScanner(file)
		for scanner.Scan() {
			allLines = append(allLines, scanner.Text())
		}
		if err := scanner.Err(); err != nil {
			fmt.Printf("Error reading file: %v\n", err)
		return
		}

		// Split into tokens
		for _, line := range allLines {
			tokens := strings.Split(line, ",")
			for i := range tokens {
				tokens[i] = strings.TrimSpace(tokens[i])
			}

		// Process each token
			for _, token := range tokens {
				if len(token) < 2 {
					continue // Skip invalid tokens
				}

				// Extract rank and suit as strings
				rank := token[:len(token)-1] // All except the last character
				suit := token[len(token)-1:] // Last character

				// Create a new Card using the constructor
				card, err := NewCard(rank, suit)
				if err != nil {
					fmt.Println("Unknown Card")
				}
				// Check for duplicates
				checkForDuplicates(card, dealtCards)

				// Add the card to the current hand
				hands[currentHand].AddCard(card)

				// Move to the next hand after 5 cards
				if hands[currentHand].cardCount() == 5 {
					currentHand++
				}
			}

		}

		fmt.Println("\n*** Here are the six hands: ")
                for _, hand := range hands {
                        fmt.Println(hand)
                }
                for _, hand := range hands {
                        hand.AssignType()
                }

	} else {

// ===================== Random Hands =======================

		fmt.Println("*** Using Random Deck ***\n")
		// Create a new deck
		deck := NewDeck()

		// Shuffle the deck
		deck.Shuffle()


		// Print the initial Deck
		fmt.Println(deck)

		// Deal 5 cards to each hand
		for i := 0; i < 5; i++ {
			for j := 0; j < 6; j++ {
				card, err := deck.DealCard()
				if err != nil {
					fmt.Println("Error dealing card:", err)
					return
				}
				hands[j].AddCard(card)
			}
		}

		// Display each hand and its sorted cards
		fmt.Println("\n*** Here are the six hands: ")
		for _, hand := range hands {
			fmt.Println(hand)
		}
		for _, hand := range hands {
			hand.AssignType()
		}
		fmt.Println("\n*** Here is what remains in the deck:")
		fmt.Println(deck)
	}
	fmt.Println("\n--- WINNING HAND ORDER ---")
	SortHands(hands)
	for _, hand := range hands {
		fmt.Println(hand)
	}

}

File: CSC330/poker/Perl/Poker.pl
#!/usr/bin/perl
use warnings;

use lib '.';  # Assuming Card.pm, Deck.pm, and Hand.pm are in the same directory
use Card;
use Deck;
use Hand;
use HandIdentifier;
use HandSorter;
# ================ Helper ===================
# Initialize an array to track dealt cards (52 cards: 13 ranks * 4 suits)
my @dealt_cards = (0) x 52;  # 0 means card has not been dealt

# A helper subroutine to check for duplicates
sub check_for_duplicates {
    my ($card) = @_;
    
    # Use the card's rank and suit integer values directly
    my $rank_value = $card->{rank};  # Rank is already an integer
    my $suit_value = $card->{suit};  # Suit is already an integer
    
    # Calculate unique card ID (rank * suit combination)
    my $card_id = ($rank_value - 1) * 4 + ($suit_value - 1);  # Offset rank and suit by 1
    
    # Check if the card has already been dealt
    if ($dealt_cards[$card_id]) {
	my $card_str = $card->to_string();
        die "Error: Duplicate card detected $card_str. Program will stop.\n";
    } else {
        # Mark the card as dealt
        $dealt_cards[$card_id] = 1;
    }
}

# ================ Main Logic ===============
# Print Title
print "*** P O K E R  H A N D  A N A L Y Z E R ***\n\n\n";
# Create an array to store the hands
my @hands;

# Initialize six empty hands
for my $i (1 .. 6) {
    my $hand = Hand->new();
    push @hands, $hand;
}

my $num_args = scalar @ARGV;
if ($num_args > 0) {
    print "*** USING TEST DECK ***\n\n";
    my $filename = $ARGV[0] or die "Need to get file name on the  command line\n";
    print "*** File: " . $filename . "\n";
    # Use the filename
    open(DATA, "<$filename") or die "Couldn't open file $filename, $!";

    #the next line puts all the lines from the text file into an array called @all_lines
    my @all_lines = <DATA>;
    my $current_hand = 0;

    foreach my $line (@all_lines) {
        my @tokens = split(',', $line);
        chomp(@tokens);  # Remove newline characters from tokens

        # Trim spaces from each token
        foreach my $token (@tokens) {
            $token =~ s/^\s+|\s+$//g;  # Remove leading and trailing spaces
	    my $rank = substr($token, 0, -1);
	    my $suit = substr($token, -1);

	    my $card = Card->new($rank, $suit);
	    check_for_duplicates($card);
	    $hands[$current_hand]->add_card($card);
	    if ($hands[$current_hand]->card_count() == 5) {
	        $current_hand++;	
            }
        }
		
        # Debug print to check the trimmed tokens
        #print "Tokens: ", join('|', @tokens), "\n";

    }    
    close(DATA);

    # Print each hand
    print "*** Here are the six hands...\n";
    for my $i (0 .. $#hands) {
        print $hands[$i]->to_string() . "\n";
    }

} else {
    print "*** USING RANDOMIZED DECK OF CARDS ***\n\n";
    my $deck = Deck->new();
    $deck->shuffle();
    print "*** Shuffled 52 card deck:\n";
    $deck->print_deck();  
    # Deal five cards to each hand
    for my $card_number (1 .. 5) {
        for my $hand (@hands) {
            my $card = $deck->deal_card();
            $hand->add_card($card);
        }
    }
    
    # Print each hand
    print "*** Here are the six hands...\n";
    for my $i (0 .. $#hands) {
        print $hands[$i]->to_string() . "\n";
    }

    print "\n*** Here is what remains in the deck...\n";
    $deck->print_deck();
}


for my $hand (@hands) {
    HandIdentifier->assign_type($hand);
}

HandSorter::sort_hands(\@hands);

print "\n--- WINNING HAND ORDER ---\n";

for my $i (0 .. $#hands) {
    print $hands[$i]->to_string() . "- " . $hands[$i]->type_to_string() . "\n";
}

